### PR TITLE
Refactor sincos API

### PR DESCRIPTION
**Pull request by steff456**
## Description

> What is the purpose of this pull request?

This pull request:

- Refactors `sincos` package to support assign when the result needs to be overwrite a given array

## Related Issues

> Does this pull request have any related issues?

This PR is part of https://github.com/stdlib-js/todo/issues/1435

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/sincos/README.md [hunk 1]
```diff
@@ -45,14 +45,16 @@ v = sincos( -3.141592653589793/6.0 );
 // returns [ ~-0.5, ~0.866 ]
 ```
 
-By default, the function returns the sine and cosine as a two-element `array`. To avoid extra memory allocation, the function supports providing an output (destination) object.
+#### sincos.assign( x, out, stride, offset )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sin] and [cosine][@stdlib/math/base/special/cos] of a `number` (in radians), and assigns results to a provided output array.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/README.md]:
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sin] and [cosine][@stdlib/math/base/special/cos] of a `number` (in radians) and assigns results to a provided output array.
```

File: lib/node_modules/@stdlib/math/base/special/sincos/benchmark/benchmark.js [hunk 2]
```diff
@@ -74,7 +74,7 @@ bench( pkg+'::separate-evaluation', function benchmark( b ) {
 	b.end();
 });
 
-bench( pkg+'::in-place', function benchmark( b ) {
+bench( pkg+'::assign', function benchmark( b ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/benchmark/benchmark.js]:
```suggestion
bench( pkg+':assign', function benchmark( b ) {
```

File: lib/node_modules/@stdlib/math/base/special/sincos/docs/repl.txt [hunk 3]
```diff
@@ -26,12 +23,37 @@
     > y = {{alias}}( NaN )
     [ NaN, NaN ]
 
+
+{{alias}}( x, out, stride, offset )
+    Simultaneously computes the sine and cosine of a number, and assigns
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/docs/repl.txt]:
```suggestion
    Simultaneously computes the sine and cosine of a number and assigns
```

File: lib/node_modules/@stdlib/math/base/special/sincos/docs/types/index.d.ts [hunk 4]
```diff
@@ -22,25 +22,53 @@
...
+	* var v = sincos( NaN );
+	* // returns [ NaN, NaN ]
+	*/
+	( x: number ): Array<number>;
+
+	/**
+	* Simultaneously computes the sine and cosine of a number.
+	*
+	* @param x - input value (in radians)
+	* @param out - destination array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/docs/types/index.d.ts]:
```suggestion
	* @param out - output array
```

File: lib/node_modules/@stdlib/math/base/special/sincos/docs/types/test.ts [hunk 5]
```diff
@@ -19,26 +19,16 @@
 /// <reference types="@stdlib/types"/>
 
 import sincos = require( './index' );
-import { Collection } from '@stdlib/types/object';
 
 
 // TESTS //
 
 // The function returns a collection...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/docs/types/test.ts]:
```suggestion
// The function returns an array of numbers...
```

File: lib/node_modules/@stdlib/math/base/special/sincos/lib/assign.js [hunk 6]
```diff
@@ -60,7 +60,7 @@ var Y = [ 0.0, 0.0 ];
 // MAIN //
 
 /**
-* Simultaneously computes the sine and cosine of a number.
+* Simultaneously computes the sine and cosine of a number, and assigns results to a provided output array.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/lib/assign.js]:
```suggestion
* Simultaneously computes the sine and cosine of a number and assigns results to a provided output array.
```

File: lib/node_modules/@stdlib/math/base/special/sincos/lib/assign.js [hunk 7]
```diff
@@ -79,27 +79,29 @@ var Y = [ 0.0, 0.0 ];
 *
 *
 * @private
-* @param {(Array|TypedArray|Object)} out - destination array
 * @param {number} x - input value (in radians)
-* @returns {(Array|TypedArray|Object)} sine and cosine
+* @param {Collection} out - destination array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/lib/assign.js]:
```suggestion
* @param {Collection} out - output array
```

File: lib/node_modules/@stdlib/math/base/special/sincos/lib/kernel_sincos.js [hunk 8]
```diff
@@ -55,12 +55,14 @@ var C6 = -1.13596475577881948265e-11; // 0xBDA8FAE9, 0xBE8838D4
 * Computes the sine and cosine on \\( \approx \[-\pi/4, \pi/4\] \\) (except for \\(-0\\)), where \\( \pi/4 \approx 0.7854 \\).
 *
 * @private
-* @param {(Array|TypedArray|Object)} out - destination array
 * @param {number} x - input value (in radians, assumed to be bounded by `~π/4` in magnitude)
 * @param {number} y - tail of `x`
-* @returns {(Array|TypedArray|Object)} sine and cosine
+* @param {Collection} out - destination array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/lib/kernel_sincos.js]:
```suggestion
* @param {Collection} out - output array
```

File: lib/node_modules/@stdlib/math/base/special/sincos/lib/main.js [hunk 9]
```diff
@@ -20,17 +20,16 @@
 
 // MODULES //
 
-var computeSincos = require( './sincos.js' );
+var computeSincos = require( './assign.js' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/lib/main.js]:
```suggestion
var assign = require( './assign.js' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/lib/main.js [hunk 10]
```diff
@@ -47,23 +46,9 @@ var computeSincos = require( './sincos.js' );
...
-* var bool = ( v === out );
-* // returns true
 */
-function sincos( out, x ) {
-	if ( arguments.length === 1 ) {
-		return computeSincos( [ 0.0, 0.0 ], out );
-	}
-	return computeSincos( out, x );
+function sincos( x ) {
+	return computeSincos( x, [ 0.0, 0.0 ], 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/lib/main.js]:
```suggestion
	return assign( x, [ 0.0, 0.0 ], 1, 0 );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.equal( typeof sincos, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function computes in-place the sine and cosine (for -256*pi < x < 0)', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
tape( 'the function computes the sine and cosine (for -256*pi < x < 0)', function test( t ) {
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided a `NaN`', function test( t ) {
+	var v;
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	v = sincos( NaN, z, 1, 0 );
+	t.equal( v, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
	t.equal( v, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided `+infinity`', function test( t ) {
+	var v;
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	v = sincos( PINF, z, 1, 0 );
+	t.equal( v, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
	t.equal( v, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = hugePositive.x;
+	sine = hugePositive.sine;
+	cosine = hugePositive.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = hugeNegative.x;
+	sine = hugeNegative.sine;
+	cosine = hugeNegative.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = largePositive.x;
+	sine = largePositive.sine;
+	cosine = largePositive.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = largeNegative.x;
+	sine = largeNegative.sine;
+	cosine = largeNegative.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = mediumPositive.x;
+	sine = mediumPositive.sine;
+	cosine = mediumPositive.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	x = mediumNegative.x;
+	sine = mediumNegative.sine;
+	cosine = mediumNegative.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+	var i;
+
+	z = [ 0.0, 0.0 ];
+	x = mediumNegative.x;
+	sine = mediumNegative.sine;
+	cosine = mediumNegative.cosine;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = sincos( x[i], z, 1, 0 );
+		t.equal( y, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
		t.equal( y, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 21]
```diff
@@ -0,0 +1,383 @@
...
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided `-infinity`', function test( t ) {
+	var v;
+	var z;
+
+	z = [ 0.0, 0.0 ];
+	v = sincos( NINF, z, 1, 0 );
+	t.equal( v, z, 'returns a reference to out' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
	t.equal( v, z, 'returns output array' );
```

File: lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+});
+
+tape( 'the function supports providing an output typed array', function test( t ) {
+	var parts;
+	var out;
+
+	out = new Float64Array( 2 );
+	parts = sincos( 1.0, out, 1, 0 );
+
+	t.strictEqual( parts, out, 'returns output typed array' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sincos/test/test.assign.js]:
```suggestion
	t.strictEqual( parts, out, 'returns output array' );
```

**Review by kgryte — CHANGES_REQUESTED** [review]

**Review by kgryte — APPROVED** [review]

**Comment by kgryte** [comment]
LGTM. Thanks, @steff456!

**Comment by kgryte** [comment]
Test failures do not seem to originate from the tests in this PR. Instead, they are coming from `besselj0`. These tests seem to be failing due to flushing subnormals to zero. Locally, `besselj0` tests pass for me.

**Comment by kgryte** [comment]
**Update**: discovered that test failures did originate from this PR. Bessel function implementations depend on `@stdlib/math/base/special/sincos`, and, in particular, the implementations depended on the refactored assignment API. However, this PR did not update dependents based on the changes in this PR.

@steff456 When performing these types of refactorings, we need to ensure that all downstream dependents are migrated to the new APIs.

Test failures should be addressed in https://github.com/stdlib-js/stdlib/commit/f2d99e153034f5953dfda019a81b4c3f81c7854b and https://github.com/stdlib-js/stdlib/commit/4e77dbfb3dfbf644b320364972b88875d33aee1c.

**Comment by steff456** [comment]
Thanks for fixing it so quickly! I'll take more care of reviewing dependents for my next PRs 😅

**Pull request merged**